hudi-utilities/src/main/java/org/apache/hudi/utilities/streamer/StreamSync.java

```diff
@@ -584,10 +605,16 @@ private Pair<SchemaProvider, Pair<String, JavaRDD<HoodieRecord>>> fetchFromSourc
                 (SchemaProvider) new DelegatingSchemaProvider(props, hoodieSparkContext.jsc(), dataAndCheckpoint.getSchemaProvider(),
                     new SimpleSchemaProvider(hoodieSparkContext.jsc(), targetSchema, props)))
             .orElse(dataAndCheckpoint.getSchemaProvider());
+        if (rowBulkInsert) {
+          return checkEmpty(new InputBatch(transformed, checkpointStr, schemaProvider), resumeCheckpointStr);
```

Comment:
isn't the 2nd arg, resumeCheckpointStr to InoputBach constr

Reply:
I feel the checkEmpty and return at 3 places is not elegant. 
can we introduce a local variable which will hold the InputBatch and eventually return it and we can do checkEmpty there. 


For eg:
```
      if (this.userProvidedSchemaProvider != null && this.userProvidedSchemaProvider.getTargetSchema() != null) {
        if (rowBulkInsert) {
          inputBatchRowWriterToReturn = new InputBatch(transformed, checkpointStr, this.userProvidedSchemaProvider), resumeCheckpointStr);
        } else {

          // If the target schema is specified through Avro schema,
          // pass in the schema for the Row-to-Avro conversion
          // to avoid nullability mismatch between Avro schema and Row schema
          if (errorTableWriter.isPresent()
              && props.getBoolean(HoodieErrorTableConfig.ERROR_ENABLE_VALIDATE_TARGET_SCHEMA.key(),
              HoodieErrorTableConfig.ERROR_ENABLE_VALIDATE_TARGET_SCHEMA.defaultValue())) {
            // If the above conditions are met, trigger error events for the rows whose conversion to
            // avro records fails.
            avroRDDOptional = transformed.map(
                rowDataset -> {
                  Tuple2<RDD<GenericRecord>, RDD<String>> safeCreateRDDs = HoodieSparkUtils.safeCreateRDD(rowDataset,
                      HOODIE_RECORD_STRUCT_NAME, HOODIE_RECORD_NAMESPACE, reconcileSchema,
                      Option.of(this.userProvidedSchemaProvider.getTargetSchema()));
                  errorTableWriter.get().addErrorEvents(safeCreateRDDs._2().toJavaRDD()
                      .map(evStr -> new ErrorEvent<>(evStr,
                          ErrorEvent.ErrorReason.AVRO_DESERIALIZATION_FAILURE)));
                  return safeCreateRDDs._1.toJavaRDD();
                });
          } else {
            avroRDDOptional = transformed.map(
                rowDataset -> getTransformedRDD(rowDataset, reconcileSchema, this.userProvidedSchemaProvider.getTargetSchema()));
          }
          schemaProvider = this.userProvidedSchemaProvider;
        }
      } else {
        Option<Schema> latestTableSchemaOpt = UtilHelpers.getLatestTableSchema(hoodieSparkContext.jsc(), fs, cfg.targetBasePath);
        // Deduce proper target (writer's) schema for the transformed dataset, reconciling its
        // schema w/ the table's one
        Option<Schema> targetSchemaOpt = transformed.map(df -> {
          Schema sourceSchema = AvroConversionUtils.convertStructTypeToAvroSchema(df.schema(),
              latestTableSchemaOpt.map(Schema::getFullName).orElse(getAvroRecordQualifiedName(cfg.targetTableName)));
          // Target (writer's) schema is determined based on the incoming source schema
          // and existing table's one, reconciling the two (if necessary) based on configuration
          return HoodieSparkSqlWriter.deduceWriterSchema(
              sourceSchema,
              HoodieConversionUtils.<Schema>toScalaOption(latestTableSchemaOpt),
              HoodieConversionUtils.<InternalSchema>toScalaOption(Option.empty()),
              HoodieConversionUtils.fromProperties(props));
        });
        // Override schema provider with the reconciled target schema
        schemaProvider = targetSchemaOpt.map(targetSchema ->
                (SchemaProvider) new DelegatingSchemaProvider(props, hoodieSparkContext.jsc(), dataAndCheckpoint.getSchemaProvider(),
                    new SimpleSchemaProvider(hoodieSparkContext.jsc(), targetSchema, props)))
            .orElse(dataAndCheckpoint.getSchemaProvider());
        if (rowBulkInsert) {
          inputBatchRowWriterToReturn = new InputBatch(transformed, checkpointStr, schemaProvider);
        } else {
          // Rewrite transformed records into the expected target schema
          avroRDDOptional = transformed.map(t -> getTransformedRDD(t, reconcileSchema, schemaProvider.getTargetSchema()));
        }
      }
    } else {
      if (rowBulkInsert) {
        inputBatchRowWriterToReturn = formatAdapter.fetchNewDataInRowFormat(resumeCheckpointStr, cfg.sourceLimit);
      } else {
        // Pull the data from the source & prepare the write
        InputBatch<JavaRDD<GenericRecord>> dataAndCheckpoint =
            formatAdapter.fetchNewDataInAvroFormat(resumeCheckpointStr, cfg.sourceLimit);
        avroRDDOptional = dataAndCheckpoint.getBatch();
        checkpointStr = dataAndCheckpoint.getCheckpointForNextBatch();
        schemaProvider = dataAndCheckpoint.getSchemaProvider();
      }
    }
```


After this, we could check checkEmpty and return right away for row writer use-case. 
but for avro we might continue further on a need basis. 

also, checkEmpty can be generalized or funcationlized so that we can re-use for both avro and spark rows

Reply:
let me know what do you think of the suggestion. happy to jam